[Hang Tian] iP

HT-T opened the pull request.

HT-T committed:
Add level-0

HT-T committed:
Level 1 commit @ 26Aug - HT

HT-T committed:
Level-2 26Aug HT

HT-T committed:
Level-3 Aug 26 HT

HT-T committed:
A-CodingStandard 26Aug HT

HT-T committed:
Aug 29; for PR

HT-T committed:
Lvl 4 Sep 1

HT-T committed:
A-CodeQuality

HT-T committed:
fixed bugs

JordanKwua commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+        protected String taskType;
+        protected String taskDate;
+    
+        public Task(String description,boolean isDone, String taskType, String taskDate) { // initiating an instance
+            this.description = description;
+            this.isDone = isDone;
+            this.taskType = taskType; // todo,ddl,event
```
I think you should standardise comments to be indented relative to their position in the code?
e.g.
while (true) {
// Do something
something();
}

cheehengk replied:
Should use inheritance for task types instead of a string variable since each type has methods and variables unique to them. Also to improve flexibility in future development

JordanKwua commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+
+        }
+
+        public String getTaskType(){
+            return (this.taskType);
+        }
+    
+        public String getStatusIcon() {
+            return (isDone ? "X" : " "); // mark done task with X
+        }
+    
+        public void markAsDone(){
+            this.isDone = true;
+        }
+
+        public void markAsNotDone(){
+            this.isDone = false;
+        }
+
+        public String getDescription(){
+            return this.description;
+
+        }
+
+        public String getDate(){
+            return this.taskDate;
+    }
+}
```
Maybe you can consider having a space after the closing bracket for the method instead?
eg.
public String getDate() {
//Insert code here
}

JordanKwua commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
         System.out.println("Hello from\n" + logo);
+        System.out.println("Hello! I'm Duke_HTT. \n What can I do for you?");
+        
+        // Take user inputt
```
Theres a spelling error here ~

JordanKwua commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+        // Processing input - 1st time
+        Vector<Task> tasks = new Vector<>();
+        
+        if (inData.equals("bye")){
+            System.out.println(" \n Bye. Hope to see you again soon!");
+        } else if (inData.equals("list")){
+            for (Task task:tasks){
+                System.out.println("Here are the tasks in your list: \n" 
+                                    + (tasks.indexOf(task) + 1) 
+                                    + "." + "[" + task.getStatusIcon() + "] " 
+                                    + task.getDescription());   
+            }
+        } else{
+            Task t = new Task(inData);
+            tasks.add(t);
+            System.out.println("added: " + t.getDescription());
+        }
```
Should follow the coding standard for if-else class of statements
eg.
if (condition) {
statements;
}

cheehengk replied:
This behaviour is repeated below in the while loop, consider removing it as it is currently obsolete and causes confusion

Brominne commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+import java.util.Vector;
+
 public class Duke {
+    public class Task {
```
create Task as separate class

Brominne commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+        }
+    
+        public String getStatusIcon() {
+            return (isDone ? "X" : " "); // mark done task with X
```
use this.isDone for better clarity

Brominne commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+
+        public String getDate(){
+            return this.taskDate;
+    }
```
maybe you want to have a toString method

JordanKwua commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+        }
+
+        // Processing input - loop
+        Boolean isBye = false;
```
Should define the boolean type as per follows (b is in small letters)
eg.
boolean isBye = false;

Brominne commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+
+            inData = scan.nextLine();
+            
+            if (inData.equals("bye")){
```
magic string "bye", define as constant

Brominne commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+            if (inData.equals("bye")){
+                isBye = true;
+                System.out.println(" \n Bye. Hope to see you again soon!");
+            } else if (inData.equals("list")){
```
magic string list

cheehengk replied:
Might want to consider using constant names such as INPUT_LIST instead of "list"

Brominne commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+                                        + "." + "[" + task.getStatusIcon() + "] " 
+                                        + task.getDescription());
+                }
+            } else if (inData.contains("mark") && !inData.contains("unmark")){
```
magic string mark, unmark

Brominne commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+                Task task = tasks.get(inDataIndex);
+                System.out.println("OK, I've marked this task as not done yet: \n" + "[" 
+                + task.getStatusIcon() + "] " + task.getDescription());
+            } else if (inData.substring(0,4).equals("todo")){
```
magic string todo

Brominne commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+                + t.getDescription() + "]" + "[" + t.getStatusIcon() + "] "
+                + t.getDescription());
+                System.out.println("Now you have %2d tasks in the list.".formatted(tasks.size()));
+            } else if (inData.substring(0,8).equals("deadline")){
```
magic string, magic number

cheehengk replied:
Same issue as above regarding numbers that are not clear in what they mean

Brominne commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+                + t.getDescription() + "]" + "[" + t.getStatusIcon() + "] "
+                + t.getDescription() + "by: " + t.getDate());
+                System.out.println("Now you have %2d tasks in the list.".formatted(tasks.size()));
+            } else if (inData.substring(0,5).equals("event")){
```
magic number, magic string

JordanKwua reviewed (commented):
Overall good job as the code works, just some minor indentation issues and placement of comments. If you're using IntelliJ maybe you could use CTRL + ALT + L to auto-indent

Brominne commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+                System.out.println("Got it. I've added this task: \n" + "["
+                + t.getDescription() + "]" + "[" + t.getStatusIcon() + "] "
+                + t.getDescription() + t.getDate());
+                System.out.println("Now you have %2d tasks in the list.".formatted(tasks.size()));
```
repetitive print command in each case, try to print common text before and after the switch statement

Brominne commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+                System.out.println("OK, I've marked this task as not done yet: \n" + "[" 
+                + task.getStatusIcon() + "] " + task.getDescription());
+            } else if (inData.substring(0,4).equals("todo")){
+                Task t = new Task(inData.substring(4),false,"T");
```
create todo as subclass of task, so [T] can be an attribute of the subclass instead of a magic string here

Brominne commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+                for (Task task:tasks){
+                    System.out.println("Here are the tasks in your list: \n" 
+                                        + (tasks.indexOf(task) + 1) 
+                                        + "[" + task.getDescription() + "]"
```
a toString method for task can entail the length code from line 93 to 95

Brominne commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+        if (inData.equals("bye")){
+            System.out.println(" \n Bye. Hope to see you again soon!");
+        } else if (inData.equals("list")){
+            for (Task task:tasks){
```
nested conditional and loop. define list as a method of task class

cheehengk commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+import java.util.Scanner;
+import java.util.Vector;
+
 public class Duke {
```
Suggest a separate file for 'Task' object

cheehengk commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+                + t.getDescription());
+                System.out.println("Now you have %2d tasks in the list.".formatted(tasks.size()));
+            } else if (inData.substring(0,8).equals("deadline")){
+                int taskEndIndex = inData.indexOf("/");
+                Task t = new Task(inData.substring(9,taskEndIndex),false,"D",
+                inData.substring(taskEndIndex+1));
+                System.out.println("Got it. I've added this task: \n" + "["
+                + t.getDescription() + "]" + "[" + t.getStatusIcon() + "] "
+                + t.getDescription() + "by: " + t.getDate());
+                System.out.println("Now you have %2d tasks in the list.".formatted(tasks.size()));
```
As also seen above, this can be refactored into a separate method to make the overall code more modular

cheehengk commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+                System.out.println("Now you have %2d tasks in the list.".formatted(tasks.size()));
+            } else if (inData.substring(0,5).equals("event")){
+                int taskEndIndex = inData.indexOf("/");
+                Task t = new Task(inData.substring(5,taskEndIndex),false,"E",
```
Suggest passing properly named variables into the constructor so that it is clearer and less prone to mistakes

cheehengk commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+                System.out.println("Nice! I've marked this task as done: \n" + "[" 
+                + task.getStatusIcon() + "] " + task.getDescription());
+            } else if (inData.contains("unmark")){
+                int inDataIndex = Integer.parseInt(inData.substring(7));
```
The number '7' is ambiguous and confusing, if required, consider using constants instead

cheehengk commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
-                + "|  _ \\ _   _| | _____ \n"
-                + "| | | | | | | |/ / _ \\\n"
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
+                    + "|  _ \\ _   _| | _____ \n"
+                    + "| | | | | | | |/ / _ \\\n"
+                    + "| |_| | |_| |   <  __/\n"
+                    + "|____/ \\__,_|_|\\_\\___|\n";
         System.out.println("Hello from\n" + logo);
+        System.out.println("Hello! I'm Duke_HTT. \n What can I do for you?");
```
Might want to consider separating this out as a method to make the code cleaner and more modular, especially since a substantially large graphic is being printed.

cheehengk commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,141 @@
+                }
+            } else if (inData.contains("mark") && !inData.contains("unmark")){
+                int inDataIndex = Integer.parseInt(inData.substring(5));
+                inDataIndex--;
```
A comment here to explain the reason for -- would be helpful, especially in future reviews of the code, since a random -- might be a potential cause of bugs

cheehengk reviewed (commented):
The code looks more procedural than object oriented. It can also be more modular by making appropriate methods to allow for flexibility and better clarity.

HT-T committed:
fixed bugs

HT-T committed:
8 Sep: Level-5 and A-Exceptions done
1) implemented Task class OOP that was absent in level 4;
2) fixed functionality bugs
3) separated functions from Duke.java;
4) changed if-else to switch case;
5) changed implementation of inputs processing

HT-T committed:
branched

HT-T committed:
Level-6 function

HT-T committed:
Level-7 funct

HT-T committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/Duke.java

exetr commented on src/main/java/Tasks/Task.java:
```diff
@@ -0,0 +1,58 @@
+
+public class Task{
+    protected String taskTitle;
+    protected boolean taskDone;
```
Perhaps consider naming this boolean variable `isDone` instead, as according to the coding standard.

exetr commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,431 @@
+    private static final String FILE_DIRECTORY_LOCATION = Paths.get(".", "save").toString(); ////////
+    protected static final String DATA_FILE_NAME = "duke" + ".txt";
+    private static final String DATA_FILE_PATH = Paths.get(FILE_DIRECTORY_LOCATION, DATA_FILE_NAME).toString();//////////
+    //protected static final String DATA_FILE_PATH = "src/main/java/" + DATA_FILE_NAME;
```
Do remember to remove commented code if it is not relevant/being used anymore

exetr commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,431 @@
+        while (fileScanner.hasNext()) {
+            String nextTask = fileScanner.next();
+            String[] taskTypes = nextTask.split(SEPARATOR);
+            Boolean taskDone = null;
```
Similarly, consider renaming this variable to `isDone` to adhere to the coding standard.

exetr commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,431 @@
+    public static void processInput(){   
+        Scanner scan = new Scanner ( System.in ); 
+        String inData = scan.nextLine(); // user line of input
+        String[] inLine = inData.split(" "); // code process line of input into an array
```
You can consider specifying your delimiter as a constant variable to minimise the use of magic literals within this block of code.

exetr commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,431 @@
+        return taskMark;
+    }
+
+    public static void processInput(){   
```
It may be good to consider whether this function seems too long. Such long functions may affect code readibility, for certain aspects of the function that are repeated, you can try to abstract them out to other functions (practice SLAP).

exetr commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,431 @@
+
+                String taskMark = getTaskMark(inLine);
+                String doneMark;
+                if (tasks.get(taskIndex).getTaskDone()){
+                    doneMark = "[X]";
+                } else{
+                    doneMark = "[ ]";
+                }
```
Similarly, in such cases, you could consider replacing the output with constant variables to eliminate magic literals.

exetr commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,431 @@
+
+            case "Deadline":
+                String deadline = tasks.get(i).getTaskDeadline();
+                //String outputIndex = i + 1;
```
Try to remove commented code that is not in use anymore.

exetr commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,431 @@
+                break;
+
+            default:
+            try {
+                throw new InvalidGeneralInputException(); // catch all other exceptions
+            } catch (InvalidGeneralInputException e) {
+                System.out.println("Error - Sorry, I do not understand this input.");
+            }
```
You might want to ensure that this block of code is indented one level further as part of the default block, adhering to the coding standards.

exetr commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,431 @@
+                break;
+
+            case "delete":
+            try {
+                if ((inputTaskIndex) > tasks.size() || (inputTaskIndex) < 1) {
+                    throw new InvalidListIndexException();
+                }
+
+                String taskMark = getTaskMark(inLine);
+                String doneMark;
+                if (tasks.get(taskIndex).getTaskDone()){
+                    doneMark = "[X]";
+                } else{
+                    doneMark = "[ ]";
+                }
+                System.out.println("Noted. I have removed this task:");
+                System.out.println(taskMark + doneMark + " " +
+                        tasks.get(taskIndex).getTaskTitle());
+                System.out.println("Now you have " + tasks.size() + " tasks in the list.");
+
+                tasks.remove(taskIndex);
+            } catch (InvalidListIndexException | NumberFormatException e) {
+                System.out.println("Error - List index given invalid. Please check again.");
+            }
+            break;
```
You should consider indenting this block of code one level further, as stated in the coding standard.

exetr commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,431 @@
+            break;
+            
+            case "todo":
+            try {
+                if (inLine.length <= 1) {
+                    throw new InvalidTodoException();
+                }
+
+                tasks.add(new ToDo(inData.substring(5)));
+
+                System.out.println("Got it. I've added this task: \n" + "[T][ ] "
+                        + inData.substring(5) ); //////////magic number problem
+                System.out.println("Now you have " + tasks.size() +  " tasks in the list.");
+
+                String newEntry = parseTaskToString(tasks);
+                writeDisk(newEntry);
+            } catch (InvalidTodoException e) {
+                System.out.println("Error - Please input a description for your To do. ");
+            } catch (IOException e) {
+                System.out.println("Error when saving file: " + e.getMessage());
+            }
+            break;
```
You should consider indenting this block of code one level further, as stated in the coding standard.

exetr commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,431 @@
+                break;
+
+            case "mark":
+            try {
+                if (inputTaskIndex > tasks.size() || inputTaskIndex < 1) {
+                    throw new InvalidListIndexException();
+                }
+
+                tasks.get(taskIndex).setTaskDone(true);
+                
+                String taskMark = getTaskMark(inLine);
+                String doneMark;
+                if (tasks.get(taskIndex).getTaskDone()){
+                    doneMark = "[X]";
+                } else{
+                    doneMark = "[ ]";
+                }
+
+                System.out.println("Nice! I've marked this task as done: \n" + taskMark + doneMark +
+                        tasks.get(taskIndex).getTaskTitle());
+
+                String newEntry = parseTaskToString(tasks);
+                writeDisk(newEntry);
+            } catch (InvalidListIndexException | NumberFormatException e) {
+                System.out.println("Error - List index given invalid. Please check again.");
+            } catch (IOException e) {
+                System.out.println("Error when saving file: " + e.getMessage());
+            }
+            break;
```
You should consider indenting this block of code one level further, as stated in the coding standard, and to maintian consistency across your code.

exetr reviewed (commented):
Great work! The few comments I've added are just to ensure that you maintian consistency across your code and that it adheres to the coding standard. Although, be aware of your `Duke.java` as there are some rather long functions that may affect code readibility!

HT-T committed:
branch_level_8

HT-T committed:
no message

HT-T committed:
Delete docs directory

HT-T committed:
Update input.txt

HT-T committed:
Merge branch 'branch-A-JavaDoc'

HT-T committed:
no message

HT-T committed:
Merge pull request #1 from HT-T/branch-PR

PR request for ip task

HT-T committed:
User's Guide

HT-T committed:
Update README.md

HT-T committed:
Update README.md

HT-T committed:
Delete runtest.bat

HT-T committed:
Delete runtest.sh
